### PR TITLE
feat(issue-alerts): preview for attributes filter

### PR DESCRIPTION
**Pull request by Zandew**
Adds support for attributes filter to preview.

### CODE REVIEW

File: src/sentry/rules/conditions/event_attribute.py [hunk 1]
```diff
@@ -1,38 +1,41 @@
...
+    "exception.value": Columns.ERROR_VALUE,
+    "user.id": Columns.USER_ID,
+    "user.email": Columns.USER_EMAIL,
+    "user.username": Columns.USER_USERNAME,
+    "user.ip_address": Columns.USER_IP_ADDRESS,
+    "http.method": Columns.HTTP_METHOD,
+    "http.url": Columns.HTTP_URL,
+    "http.status_code": Columns.HTTP_STATUS_CODE,
+    "sdk.name": Columns.SDK_NAME,
+    "stacktrace.code": None,
```
Comment (Zandew) [inline src/sentry/rules/conditions/event_attribute.py:29]: I can't find the corresponding column for this attribute, the stacktrace code should be made up of `pre_context`, `context`, and `post_context`, do you know if these are queryable?

Comment (rjo100) [reply src/sentry/rules/conditions/event_attribute.py:29]: special case: we are looking for the `frame` `context_line`. see `test_stacktrace_code` in `tests/sentry/rules/conditions/test_event_attribute.py`

Comment (Zandew) [reply src/sentry/rules/conditions/event_attribute.py:29]: decided that we're not doing this at least for now

File: tests/sentry/rules/history/test_preview.py [hunk 2]
```diff
@@ -212,6 +212,25 @@ def test_tagged(self):
         results = preview(self.project, conditions, filters, "all", "all", 0)
         assert event.group not in results
 
+    def test_event_attribute(self):
+        event = self._set_up_event({"message": "message"})
```
Comment (rjo100) [inline tests/sentry/rules/history/test_preview.py:216]: change the `value` here to something other than `message`. too confusing

File: src/sentry/rules/conditions/event_attribute.py [hunk 3]
```diff
@@ -189,21 +192,14 @@ def render_label(self) -> str:
...
         return self.label.format(**data)
 
-    def passes(self, event: GroupEvent, state: EventState, **kwargs: Any) -> bool:
-        attr = self.get_option("attribute")
+    def _passes(self, attribute_values: Sequence[Any]) -> bool:
         match = self.get_option("match")
         value = self.get_option("value")
 
-        if not (attr and match and value):
+        if not (match and (value or match in (MatchType.IS_SET, MatchType.NOT_SET))):
```
Comment (rjo100) [inline src/sentry/rules/conditions/event_attribute.py:199]: is `value` always `None` with the `is_set` and `is_not_set` matchers? maybe add a second test for one of them.

File: src/sentry/rules/conditions/event_attribute.py [hunk 4]
```diff
@@ -262,3 +258,38 @@ def passes(self, event: GroupEvent, state: EventState, **kwargs: Any) -> bool:
             return not attribute_values
 
         raise RuntimeError("Invalid Match")
+
+    def passes(self, event: GroupEvent, state: EventState, **kwargs: Any) -> bool:
+        attr = self.get_option("attribute")
+        try:
+            attribute_values = self._get_attribute_values(event, attr.lower())
```
Comment (rjo100) [inline src/sentry/rules/conditions/event_attribute.py:265]: are we losing the logic for if `attr` is empty or None?

File: src/sentry/rules/conditions/event_attribute.py (hunk 3 above)
Comment (rjo100) [reply src/sentry/rules/conditions/event_attribute.py:199]: I would maybe separate out the logic here to make it a bit cleaner to read//less likely to get misinterpreted
`not ((match and value) or (match in (MatchType.IS_SET, MatchType.NOT_SET)))`

File: src/sentry/rules/conditions/event_attribute.py (hunk 4 above)
Comment (Zandew) [reply src/sentry/rules/conditions/event_attribute.py:265]: I don't think its actually possible for `attr` to be none, it should have to be one of the choices? But if it is, i think it should still fail accordingly

Comment (rjo100) [reply src/sentry/rules/conditions/event_attribute.py:265]: yeah but wouldn't it throw a `TypeError` or something? not a `KeyError` (should happen during the `attr.lower()` call

Comment (rjo100) [reply src/sentry/rules/conditions/event_attribute.py:265]: that's only if it's `None` and not empty

Comment (Zandew) [reply src/sentry/rules/conditions/event_attribute.py:265]: yeah i if its none then it'll throw an `AttributeError` which we don't catch, and if its empty then `_get_attribute_values` returns an empty list. But i think attr has to be in `ATTR_CHOICES` and i think some validation is probably done before this

Comment (rjo100) [reply src/sentry/rules/conditions/event_attribute.py:265]: you can never really guarantee a frontend variable is sent successfully to the backend due to manual customer/hacker requests and just like faulty frontend stuff (like adblockers or js not properly loading). I'd either add a default of `''` to the `get_option` call or explicitly catch the exception that `.lower()` throws on a `None` type obejct.

without knowing too much about why the logic was there previously, in the old code we checked to see if it was empty/None and now we don't, which feels like a regression

**Review by rjo100 — APPROVED** [review]

**Pull request merged**